Counter meter to send its data as Stackdriver's CUMULATIVE MetricKind

tmarszal opened the pull request:
https://github.com/micrometer-metrics/micrometer/issues/2231

tmarszal commented:
I am working on unit tests for it, but would like to know your opinion on this early on.

shakuzen commented:
This generally sounds like a good improvement. `Timer`s and `Summary`s also have a count that should probably also be changed to the CUMULATIVE type. 
About the interval start and end times, previously it kept all of the end times the same for a given publish interval, but it looks like with the change it will be potentially different for each time series. I'm not sure which is better or what effect this has in how things are displayed or calculated in Stackdriver, but it is something to consider.

tmarszal commented:
You are right, Timer's and Summary's counts should also be CUMULATIVE. I went through all the meter types and also found FunctionCount and added CUMULATIVE there.

Regarding interval times, I don't see any drawbacks of them being different between timeseries, but it would be better to keep their current behavior, so I modified the code.

tmarszal commented:
I am aware my change broke some tests - would try to fix it this afternoon.

shakuzen commented on implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java:
```diff
@@ -325,11 +340,23 @@ TimeSeries createTimeSeries(Meter meter, Distribution distribution) {
             String metricType = metricType(id, statistic);
 
+            if (!metricKind.equals(backendMetricKind)) {
+                logger.info("MetricKind configured in Stackdriver ({}) is not matching the one expected by Micrometer for that kind of Meter ({})." +
```
Will this be logged every publish for every time series that is mismatched? That sounds overly verbose if so. Ideally it would be logged once on the first publish and then not again. Rather than a log per time series, one per publish would also be better from a verbosity perspective.

YuliaFrishko replied:
@shakuzen you are right it can be very verbose for the user. But this log is important. And we can't move it to the publish() method level without adding a lot of complexity, which is not justified in this case. 
So I changed the logging level to 'trace'.

shakuzen commented on implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java:
```diff
@@ -282,16 +284,29 @@ public StackdriverMeterRegistry build() {
+
+        Batch() {
+            long wallTime = clock.wallTime();
+            startTime = buildTimestamp(wallTime - config.step().toMillis());
```
If this isn't a big issue, we could improve this in a follow-up change rather than block merging this change.

YuliaFrishko replied:
In this case we need to save endDate value of the previous period for CUMULATIVE and DELTA metrics separately. And then use this value + 1 milisec as the startDate for the next period. I want to create map to store these values. I’m working on it now, I’m going to test it and will push by evening.

YuliaFrishko replied:
Added the changes described above. Created field `previousBatchEndDate` and save there the last end date of previous batch. Use this field for initialising startDate `startDate = previousBatchEndDate + 1 milisec`
And situation with a final push on shutdown will be covered in this case correctly

shakuzen reviewed (commented):
I've left some feedback. In addition, maybe we should consider using cumulative meters instead of `Step*` types, which seem to go against the goal of providing cumulative types to Stackdriver. With using Step variants, the cumulative count effectively resets every publish.

pivotal-issuemaster commented:
@tmarszal Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2232)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2232) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@tmarszal Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2232)!

tmarszal commented:
> maybe we should consider using cumulative meters instead of `Step*` types, which seem to go against the goal of providing cumulative types to Stackdriver. With using Step variants, the cumulative count effectively resets every publish.

From Stackdriver point of view the difference between micrometer's Step and Cumulative is just how we play with the startTime. The recorded result is a reflection of this and for the needs of SLOs, it brings the same information. As no one raised it as an issue before, I would not change it now, but happy to change it in the future, if needed.

To illustrate startTime play. Let's assume we are sending tuples of (startTime, endTime, value):
|  | t1 | t2 | t3 | t4 | t5 | t6 |
|-|-|-|-|-|-|-|
| Step | (t0, t1, 1) | (t1+1ms, t2, 0) | (t2+1ms, t3, 1) | (t3+1ms, t4, 1) | (t4+1ms, t5, 2) | (t5+1ms, t6, 3) |
| Cumulative | (t0, t1, 1) | (t0, t2, 1) | (t0, t3, 2) | (t0, t4, 3) | (t0, t5, 5) | (t0, t6, 8) |

pebo commented:
I noticed that MicroMeter StackDriver logs counters as GAUGE / double rather than CUMULATIVE / int64 and this PR seems to address that! Will this PR be included in 1.8 or are there other concerns?

tmarszal commented:
Hi again, is there anything that I can do help this PR?

shakuzen commented:
Sorry this keeps slipping. While this still generally sounds like the direction we want to move, it sounds like this would be a breaking change for users currently using the count metrics as is. If this is breaking, I think it will be better to introduce it gradually with a configuration option to choose the previous behavior or this proposed behavior, defaulting to the previous behavior initially. I started trying to do some testing with this now that I have a proper GCP account setup, but I needed to delete the previous metric descriptor in order to write the new cumulative one, and I could not figure out how to do that from the web UI.

> From Stackdriver point of view the difference between micrometer's Step and Cumulative is just how we play with the startTime.

While I can see how the data works out that way, it's really a delta we are sending rather than a cumulative value - hence my confusion and hesitation to call it a cumulative metric kind. [I see that there is also a delta metric kind](https://cloud.google.com/monitoring/api/v3/kinds-and-types). Would it not be better to use the delta kind instead?

> Regarding interval times

With #1218, the push timing is better aligned with the Step- meters' interval. It would be most accurate if we use the step interval the values will actually be from, rather than the publish time. Values in a publish are from the immediately prior interval.

tmarszal commented:
> If this is breaking, I think it will be better to introduce it gradually with a configuration option to choose the previous behavior or this proposed behavior, defaulting to the previous behavior initially.

I was aware this is the change in behavior. How we implemented it is that the Registry checks if the metric descriptor already exists. If it does, it honors the type that's already there. If it is about to create a new metric descriptor, it uses the type that we decided is best (so Cumulative for now; but it can be changed). But if configuration option is more aligned with what was used before in such cases, I am happy to change it :) Is that the case?

> I started trying to do some testing with this now that I have a proper GCP account setup, but I needed to delete the previous metric descriptor in order to write the new cumulative one, and I could not figure out how to do that from the web UI.

I am afraid it can't be done from UI. The best option is to use API Explorer: https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors/delete. Following on the previous point, if you start your application when the previous metric descriptor exist, micrometer should still write to it.

> Would it not be better to use the delta kind instead?

I’d like to reconsider. It’s likely that you are right and Delta is a better fit for this.

> #1218

I will reconsider how we use times :)

tmarszal commented:
To set timing expectations here, I am heading for 2 weeks of holidays, so the earliest I would pick it up is late August/early September.
@YuliaFrishko if you'd like to move this forward before that, I am more than happy.

shakuzen commented:
> if you start your application when the previous metric descriptor exist, micrometer should still write to it.

When I tried, I got the following error:

```
com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: One or more TimeSeries could not be written: Value type for metric custom.googleapis.com/http/server/requests/count must be INT64, but is DOUBLE.: timeSeries[2]
```

It's probably a mistake that we ever let it default to INT64 before, since counts can be doubles, even if they most often aren't. Nonetheless, this looks like a compatibility issue users will face with things as is.

If we can make it work without a configuration option, users would still have to delete the metadata to get the new behavior. It would also yield different behavior compared to old metrics for any new metrics introduced after upgrading to a version with this. Without a configuration option, the user has no way to choose to keep things consistent except deleting the old metadata to push everything to the new way. If we get enough user feedback, we can remove the configuration option in Micrometer 2.0 or at least flip the default to encourage adoption of the new metadata.

> To set timing expectations here, I am heading for 2 weeks of holidays, so the earliest I would pick it up is late August/early September.

1.8.0-M3 is scheduled for mid September, so if we can aim to get it merged for that, it would be great. If not, there's an RC1 a month after that.

YuliaFrishko commented:
> When I tried, I got the following error:
> 
> ```
> com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: One or more TimeSeries could not be written: Value type for metric custom.googleapis.com/http/server/requests/count must be INT64, but is DOUBLE.: timeSeries[2]
> ```
> 
Hi @shakuzen
I'm trying to reproduce the issue you mentioned. 
Could you provide more information on how you received this error(probably more info of the metric you used). I tried to run with existed GAUGE metric and didn't get any errors, but probably there is some tiny details i'm missing on how your metric was created

shakuzen commented:
@YuliaFrishko thank you for trying to take a look into it. Sorry I didn't provide much information on how to reproduce it. This is happening with a `Timer` meter, specifically `http.server.requests` which is used in Spring Boot instrumentation of timing HTTP server requests such as WebMVC and WebFlux endpoints. From the `http.server.requests` Timer meter, it is the time series for the `count` that is having an issue. It has been created as INT64 value type previously. With the changes from this pull request, it is trying to be sent as a DOUBLE type instead. Hopefully that helps you be able to reproduce it. It should happen with any `Timer` published on a version without these changes and then later on a version with these changes, I think.

jonatan-ivanov commented:
Fyi, there was an issue opened that covers this PR so I moved the labels and milestone there.

tmarszal commented:
> I noticed that MicroMeter StackDriver logs counters as GAUGE / double rather than CUMULATIVE / int64 and this PR seems to address that! Will this PR be included in 1.8 or are there other concerns?

@pebo I read you comments and the reason is that Counter implementation uses double itself. It's done consistently, including [examples](https://micrometer.io/docs/concepts#_counters), which make me think it is intended behavior. Do you see any issues if we start using CUMULATIVE (or DELTA) / double?

pebo commented:
> > I noticed that MicroMeter StackDriver logs counters as GAUGE / double rather than CUMULATIVE / int64 and this PR seems to address that! Will this PR be included in 1.8 or are there other concerns?
> 
> @pebo I read you comments and the reason is that Counter implementation uses double itself. It's done consistently, including [examples](https://micrometer.io/docs/concepts#_counters), which make me think it is intended behavior. Do you see any issues if we start using CUMULATIVE (or DELTA) / double?

I doubt that the metric kind and value type can be changed for existing time series but I haven't tried it, e.g. what happens if the metric kind suddenly changes from `gauge` to `cumulative` (https://cloud.google.com/monitoring/api/v3/kinds-and-types) due to a change in MicroMeter?

Currently I'm using using MicroMeter StackDriver 1.7.x  and while the model with kind gauge and value type double for discrete counters is a bit odd it actually seems to work fine.

If the changes in the PR aren't compatible with StackDriver metrics written with MicroMeter before this change I guess there should be a note in the release notes explaining that this is breaking change for existing time series?

tmarszal commented:
You are right, MetricKind and ValueType of exising metrics cannot be changed. Our current approach for backward compatibility is that the new behavior is controlled by the flag. It is false by default and unless you change it, everything works as it used to. Once set to true, it tries sending data using MetricKinds we consider better.

tmarszal commented:
Quick recap of what was done so far:
* MetricKind was changed to CUMULATIVE for: FunctionCounter, Counter, and count as a part of Timer and Summary representation
* new behavior is breaking, so the change was hidden behind a configuration flag, which is set to `false` by default
* test was added to ensure old and new behavior are working as expected
* createTimeSeries methods were reorganized to make it more obvious which meter type is creating which kind of metric
* I checked and confirmed we are recording interval time at similar moment other registries are doing so (compared to cloudwatch2)
* added a test for StepCounter to demonstrate it is resetting its value regularly

shakuzen commented:
FYI I polished the JavaDoc in c91d321
